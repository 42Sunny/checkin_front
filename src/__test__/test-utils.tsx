// test-utils.jsx
import { configureStore } from "@reduxjs/toolkit";
import { Queries, render, RenderResult } from "@testing-library/react";
import { createMemoryHistory } from "history";
import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
// Import your own reducer
import renderer from "react-test-renderer";
import appStore from "../redux/configureStore";
import rootReducer, { RootState } from "../redux/modules";
import { initialState as clusterInitState } from "../redux/modules/cluster";
import { initialState as userInitState } from "../redux/modules/user";

const makeCustomStore = ({
  cluster,
  user,
}: {
  cluster?: typeof clusterInitState;
  user?: typeof userInitState;
} = {}) => {
  return {
    clusterReducer: { ...clusterInitState, ...cluster },
    userReducer: { ...userInitState, ...user },
  };
};

const customRender = (
  ui: ReactElement,
  preloadedState: RootState = makeCustomStore(),
  { store = configureStore({ reducer: rootReducer, preloadedState }), ...renderOptions } = {},
  history = createMemoryHistory(),
): RenderResult<Queries, HTMLElement> => {
  const Wrapper: any = ({ children }: { children: ReactElement }) => {
    return (
      <Provider store={store}>
        <Router history={history}>{children}</Router>
      </Provider>
    );
  };
  return render(ui, { wrapper: Wrapper, ...renderOptions });
};

const createWithWrapper = (children: ReactElement) =>
  renderer.create(
    <Provider store={appStore}>
      <Router history={createMemoryHistory()}>{children}</Router>
    </Provider>,
  );

// re-export everything
export * from "@testing-library/react";
// override render method
export { customRender as render, createWithWrapper, makeCustomStore };
