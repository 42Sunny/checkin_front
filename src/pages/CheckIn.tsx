import { Backdrop, CircularProgress } from "@mui/material";
import React, { useCallback, useEffect, useRef, useState } from "react";
import { useHistory } from "react-router-dom";
import { UserApi } from "../api";
import ClusterStatusBoard from "../components/ClusterStatusBoard";
import ProfileCard from "../components/ProfileCard";
import TimeLogCard from "../components/TimeLogCard";
import classes from "../styles/pages/CheckInPage.module.css";
import { removeCookieValue } from "../utils/cookie";
import useCluster from "../utils/hooks/useCluster";
import useUser from "../utils/hooks/useUser";
import { formatToGeneralTime } from "../utils/time";

const getUserStatus = async () => {
  const getUserStatusRes = await UserApi.getUserStatus();
  const {
    user: { card, login, profile_image_url, state, checkin_at, checkout_at },
    cluster: { gaepo, seocho },
  } = getUserStatusRes.data;
  const cardNum = card !== null ? card : "";
  return {
    user: {
      state: state || "checkOut",
      id: login,
      cardNum,
      checkinAt: checkin_at,
      checkoutAt: checkout_at,
      profile: profile_image_url,
    },
    cluster: { gaepo, seocho },
  };
};

const getLogs = async () => {
  const today = new Date();
  const from = formatToGeneralTime(new Date(today.getFullYear(), today.getMonth(), 1));
  const to = formatToGeneralTime(new Date(today.getFullYear(), today.getMonth() + 1, 0));

  const response = await UserApi.getDailyUsage({ from, to });
  const logData = response.data.list;
  return logData.reverse();
};

const ALREADY_CHECK_IN_ERROR = "이미 체크인 되었습니다." as const;
const GENERAL_CHECK_IN_ERROR = "체크인을 처리할 수 없습니다. 관리자에게 문의해주세요." as const;
const ALREADY_CHECK_OUT_ERROR = "이미 체크아웃 되었습니다." as const;
const GENERAL_CHECK_OUT_ERROR = "체크아웃을 처리할 수 없습니다. 관리자에게 문의해주세요." as const;
const CheckIn = () => {
  const checkInCardWrapper = useRef<HTMLDivElement>(null);
  const history = useHistory();
  const {
    setUser,
    logout,
    user: { state: userState },
  } = useUser();
  const { setCurrentUserCount } = useCluster();

  const [isCardFlipped, setIsCardFlipped] = useState(false);
  const [logs, setLogs] = useState<Log[]>([]);
  const [isLoading, setIsLoading] = useState(false);

  const getUserData = useCallback(async () => {
    setIsLoading(true);
    try {
      const [getUserStatusData, getLogsData] = await Promise.all([getUserStatus(), getLogs()]);
      setUser(getUserStatusData.user);
      setCurrentUserCount(getUserStatusData.cluster);
      setLogs(getLogsData);
    } catch (e) {
      setLogs([]);
      alert("유저 정보가 올바르지 않습니다.\n 반복될 경우 관리자에게 요청해주세요");
      removeCookieValue(process.env.REACT_APP_AUTH_KEY);
      logout();
    }
    setIsLoading(false);
  }, [logout, setCurrentUserCount, setUser]);

  const handleFlip = () => {
    setIsCardFlipped((prev) => !prev);
  };

  const handleCheckIn = useCallback(
    (cardNum: string) => async (e: React.FormEvent<HTMLFormElement>) => {
      e.preventDefault();
      setIsLoading(true);
      try {
        if (userState === "checkIn") throw new Error(ALREADY_CHECK_IN_ERROR);
        const { data: checkinData } = await UserApi.postCheckIn({ cardNum });
        if (!checkinData.result) throw new Error(GENERAL_CHECK_IN_ERROR);
        history.push("/end");
        return true;
      } catch (err: any) {
        let message = GENERAL_CHECK_IN_ERROR;
        message = err?.response?.data?.message || err.message || message;
        alert(message);
        throw err;
      } finally {
        setIsLoading(false);
      }
    },
    [history, userState],
  );

  const handleCheckOut = useCallback(async () => {
    setIsLoading(true);
    try {
      if (userState === "checkOut") throw new Error(ALREADY_CHECK_OUT_ERROR);
      const { data } = await UserApi.postCheckOut();
      if (!data) throw new Error(GENERAL_CHECK_OUT_ERROR);
      history.push("/end");
    } catch (err: any) {
      let message = GENERAL_CHECK_OUT_ERROR;
      message = err?.response?.data?.message || err.message || message;
      alert(message);
      window.location.reload();
      throw err;
    } finally {
      setIsLoading(false);
    }
  }, [history, userState]);

  useEffect(() => {
    getUserData();
    return () => {
      setIsLoading(false);
    };
  }, [getUserData]);
  return (
    <>
      <Backdrop style={{ zIndex: 1 }} open={isLoading}>
        <CircularProgress size={50} color='inherit' />
      </Backdrop>
      <ClusterStatusBoard />
      <div
        ref={checkInCardWrapper}
        className={`${classes["card-wrapper"]} ${!isCardFlipped ? classes.front : classes.back}`}
      >
        <ProfileCard
          handleCheckIn={handleCheckIn}
          handleCheckOut={handleCheckOut}
          handleFlip={handleFlip}
        />
        <TimeLogCard logs={logs} handleFlip={handleFlip} />
      </div>
    </>
  );
};
export default CheckIn;
