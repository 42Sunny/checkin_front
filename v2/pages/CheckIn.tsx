import { Backdrop, CircularProgress } from "@mui/material";
import React, { useCallback, useEffect, useState } from "react";
import { useHistory } from "react-router-dom";
import { getClusterUsingInfo } from "../api/configAPI";
import { getDailyUsage, getStatus, postCheckIn, postCheckOut } from "../api/userAPI";
import ClusterStatusBoard from "../components/common/ClusterStatusBoard";
import ProfileCard from "../components/ProfileCard";
import TimeLogCard from "../components/TimeLogCard";
import classes from "../styles/pages/CheckInPage.module.css";
import { removeCookieValue } from "../utils/cookie";
import useCluster from "../utils/hooks/useCluster";
import useUser from "../utils/hooks/useUser";
import { formatToGeneralTime } from "../utils/time";

const getUserStatus = async () => {
  const response = await getStatus();
  const clusterResponse = await getClusterUsingInfo();
  const {
    user: { card, login, profile_image_url, state, checkin_at, checkout_at },
    isAdmin,
  } = response.data.payload;
  const { gaepo, seocho } = clusterResponse.data.payload;
  const cardNum = card !== null ? card : "";
  return {
    user: {
      state: state || "checkOut",
      id: login,
      cardNum,
      checkinAt: checkin_at,
      checkoutAt: checkout_at,
      profile: profile_image_url,
    },
    cluster: { gaepo, seocho },
    isAdmin,
  };
};

export const getLogs = async () => {
  const today = new Date();
  const from = formatToGeneralTime(new Date(today.getFullYear(), today.getMonth(), 1));
  const to = formatToGeneralTime(new Date(today.getFullYear(), today.getMonth() + 1, 0));
  const response = await getDailyUsage({ from, to });
  const logData = response.data.payload;
  return logData.reverse();
};

const ALREADY_CHECK_IN_ERROR = "이미 체크인 되었습니다." as const;
const GENERAL_CHECK_IN_ERROR = "체크인을 처리할 수 없습니다. 관리자에게 문의해주세요." as const;
const ALREADY_CHECK_OUT_ERROR = "이미 체크아웃 되었습니다." as const;
const GENERAL_CHECK_OUT_ERROR = "체크아웃을 처리할 수 없습니다. 관리자에게 문의해주세요." as const;

const CheckIn = () => {
  const history = useHistory();
  const {
    setUser,
    logout,
    setAuth,
    user: { state: userState },
  } = useUser();
  const { setCurrentUserCount } = useCluster();

  const [isCardFlipped, setIsCardFlipped] = useState(false);
  const [logs, setLogs] = useState<Log[]>([]);
  const [isLoading, setIsLoading] = useState(false);

  const getUserData = useCallback(async () => {
    setIsLoading(true);
    try {
      const [getUserStatusData, getLogsData] = await Promise.all([getUserStatus(), getLogs()]);
      setUser(getUserStatusData.user);
      setAuth({ isAdmin: getUserStatusData.isAdmin });
      setCurrentUserCount(getUserStatusData.cluster);
      setLogs(getLogsData);
    } catch (e) {
      setLogs([]);
      console.log(e);
      alert("유저 정보가 올바르지 않습니다.\n 반복될 경우 관리자에게 요청해주세요");
      removeCookieValue(process.env.REACT_APP_AUTH_KEY);
      logout();
    }
    setIsLoading(false);
  }, [logout, setAuth, setCurrentUserCount, setUser]);

  const handleFlip = () => {
    setIsCardFlipped((prev) => !prev);
  };

  const handleCheckIn = useCallback(
    (cardNum: string) => async (e: React.FormEvent<HTMLFormElement>) => {
      e.preventDefault();
      setIsLoading(true);
      try {
        if (userState === "checkIn") throw new Error(ALREADY_CHECK_IN_ERROR);
        const response = await postCheckIn({ cardNum });

        if (response.status !== 200) throw new Error(GENERAL_CHECK_IN_ERROR);
        history.push("/end");
        return true;
      } catch (error: any) {
        // TODO: message undefined 확인해야함.

        console.log(error);
        const message = GENERAL_CHECK_IN_ERROR;
        //  message = error.message || message;
        alert(message);
        window.location.reload();
        throw error;
      } finally {
        setIsLoading(false);
      }
    },
    [history, userState],
  );

  const handleCheckOut = useCallback(async () => {
    setIsLoading(true);
    try {
      if (userState === "checkOut") throw new Error(ALREADY_CHECK_OUT_ERROR);
      const { data } = await postCheckOut();
      if (!data) throw new Error(GENERAL_CHECK_OUT_ERROR);
      history.push("/end");
    } catch (err: any) {
      const message = GENERAL_CHECK_OUT_ERROR;
      //  message = err.message || message;
      alert(message);
      window.location.reload();
      throw err;
    } finally {
      setIsLoading(false);
    }
  }, [history, userState]);

  useEffect(() => {
    getUserData();
    return () => {
      setIsLoading(false);
    };
  }, [getUserData]);

  return (
    <>
      <Backdrop style={{ zIndex: 1 }} open={isLoading}>
        <CircularProgress size={50} color='inherit' />
      </Backdrop>
      <ClusterStatusBoard />
      <div
        className={`${classes["card-wrapper"]} ${!isCardFlipped ? classes.front : classes.back}`}
      >
        <ProfileCard
          handleCheckIn={handleCheckIn}
          handleCheckOut={handleCheckOut}
          handleFlip={handleFlip}
        />
        <TimeLogCard logs={logs} handleFlip={handleFlip} />
      </div>
    </>
  );
};
export default CheckIn;
