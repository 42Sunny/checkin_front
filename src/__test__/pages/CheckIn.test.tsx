import CheckIn from "../../pages/CheckIn";
import { render, screen, createWithWrapper } from "../test-utils";

describe("src/pages/CheckIn.tsx", () => {
  it("matches snapshot", () => {
    const component = createWithWrapper(<CheckIn />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
