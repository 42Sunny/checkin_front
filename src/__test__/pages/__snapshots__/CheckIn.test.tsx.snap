// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/CheckIn.tsx matches snapshot 1`] = `
Array [
  <div
    aria-hidden={true}
    className="MuiBackdrop-root css-sox5kk-MuiBackdrop-root"
    style={
      Object {
        "opacity": 0,
        "visibility": "hidden",
        "zIndex": 1,
      }
    }
  >
    <span
      className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorInherit css-62e83j-MuiCircularProgress-root"
      role="progressbar"
      style={
        Object {
          "height": 50,
          "width": 50,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </span>
  </div>,
  <div
    className="wrap"
  >
    <div
      className="item"
    >
      <span
        className="title"
      >
        개포
        <span
          className="count"
        >
          <span
            data-testid="개포-cadet-count"
          >
            0
             / 
            0
          </span>
        </span>
        <div
          className="circle green"
          role="status"
        />
      </span>
    </div>
    <div
      className="item"
    >
      <span
        className="title"
      >
        서초
        <span
          className="count"
        >
          <span
            data-testid="서초-cadet-count"
          >
            0
             / 
            0
          </span>
        </span>
        <div
          className="circle green"
          role="status"
        />
      </span>
    </div>
  </div>,
  <div
    className="card-wrapper front"
  >
    <div
      className="profileCard"
    >
      <img
        alt="logo"
        className="logo"
        src="42-logo-black.png"
      />
      <div
        className="util-box"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="flip-button"
          focusable="false"
          onClick={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M3 13h2v-2H3v2zm0 4h2v-2H3v2zm0-8h2V7H3v2zm4 4h14v-2H7v2zm0 4h14v-2H7v2zM7 7v2h14V7H7z"
          />
        </svg>
      </div>
      <div
        className="profile-wrapper"
      >
        <img
          alt="profile"
          className="profile"
          src="user-default.png"
        />
        <h2
          className="userId"
        >
          
        </h2>
      </div>
      <hr
        className="divider"
      />
      <div
        className="box"
      >
        <p>
           클러스터 운영시간: 
          
        </p>
        <p>
           인포데스크 점심시간 
          
        </p>
      </div>
      <form
        className="check-in-form"
        onSubmit={[Function]}
      >
        <input
          className="cardNumber"
          inputMode="decimal"
          min={1}
          onChange={[Function]}
          placeholder="카드 번호"
          step={1}
          tabIndex={0}
          type="number"
          value=""
        />
        <div
          className="checkWrap"
        >
          <label
            className="text"
          >
            <input
              checked={false}
              className="checkbox"
              onChange={[Function]}
              type="checkbox"
            />
            방역수칙에 동의하고 입장합니다.
          </label>
          <button
            className="checkDetail"
            onClick={[Function]}
            type="button"
          >
            자세히
          </button>
        </div>
        <button
          className="button green"
          color="green"
          disabled={true}
          type="submit"
        >
          CHECK IN
        </button>
      </form>
    </div>
    <div
      className="time-log-wrapper"
    >
      <div
        className="time-log-header"
      >
        <h4
          className="time-log-title"
        >
          클러스터 로그
        </h4>
        <p
          className="time-log-subtitle"
        >
          체크인 시각: 
          
        </p>
      </div>
      <div
        className="util-box"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="AccountBoxIcon"
          focusable="false"
          onClick={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M3 5v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2H5c-1.11 0-2 .9-2 2zm12 4c0 1.66-1.34 3-3 3s-3-1.34-3-3 1.34-3 3-3 3 1.34 3 3zm-9 8c0-2 4-3.1 6-3.1s6 1.1 6 3.1v1H6v-1z"
          />
        </svg>
      </div>
      <li
        className="log-data-header"
      >
        <div>
          날짜
        </div>
        <div>
          이용 시간
        </div>
      </li>
      <hr
        className="divider"
      />
      <ul
        className="log-data-wrapper"
      />
    </div>
  </div>,
]
`;
